File: src/lira/linguistics/role/linguistic_controller.py
from typing import List, Optional

from .clause_reader import ClauseReader
from .graph_processor import GraphProcessor
from .grammar_configurator import GrammarConfigurator
from .lexer import LinguisticLexer
from .phrase_reader import PhraseReader
from .prompt_tokenizer import PromptTokenizer
from .reading_context import ReadingContext
from .sentence_reader import SentenceReader
from .sequence_engine import SequenceEngine
from .token_resolver import TokenResolver
from ..data.sentence import Sentence
from ..data.tensor import LinguisticSystemPropertyTensor
from ..data.subject import Subject
from ..ui.user_prompt import UserPrompt

# DictionaryProcessor, LexicalRelationshipStore (lira.vocabulary) are
# used only as type hints here -- see graph_processor.py for why
# they're deliberately left unimported at module scope.


class LinguisticController:
    def __init__(
        self,
        dictionary_processor: "DictionaryProcessor",
        use_clause_segmentation: bool = True,
        lexical_relationships: "Optional[LexicalRelationshipStore]" = None,
    ):
        """dictionary_processor: Vocabulary owns the lexicon (Rule 17);
        Linguistics resolves tokens through it rather than keeping its
        own copy (typically Domain.vocabulary.dictionary_processor).
        lexical_relationships: plumbed through for Phase 2 (real
        morphological agreement scoring -- see linguistics/documentation/
        README.md, Not Yet Built); no Phase 1 reader consults it. Added
        now, as a one-time constructor signature change, so Domain's own
        wiring (knowledge/data/domain.py) doesn't need a second breaking
        change once Phase 2 lands."""
        self.grammar_configurator = GrammarConfigurator()
        # A typo in a rule table fails here, at construction time, not
        # mid-parse (grammar_configurator.py's own docstring on this
        # method).
        self.grammar_configurator.validate_against_vocabulary()

        self.tensor = LinguisticSystemPropertyTensor()  # persistent, canonical store for every unit's numeric fields (Rule 14)
        self.graph_processor = GraphProcessor(dictionary_processor, self.grammar_configurator, self.tensor, use_clause_segmentation)
        self.tokenizer = PromptTokenizer(self.graph_processor)
        self.lexical_relationships = lexical_relationships

        sequence_engine = SequenceEngine(self.grammar_configurator)
        token_resolver = TokenResolver(self.graph_processor)
        phrase_reader = PhraseReader(sequence_engine, self.graph_processor, self.grammar_configurator)
        clause_reader = ClauseReader(phrase_reader, sequence_engine, self.grammar_configurator)
        sentence_reader = SentenceReader(clause_reader, token_resolver, sequence_engine, self.grammar_configurator)

        # Built once, held for the controller's lifetime -- every
        # Phrase.read()/Clause.read()/Sentence.read() call reaches these
        # same shared services through this bundle (role/reading_context.py's
        # own docstring; spec 9's "Sentence.read() -> LinguisticController
        # -> shared sequencing services").
        self.reading_context = ReadingContext(
            grammar=self.grammar_configurator, sequence_engine=sequence_engine, token_resolver=token_resolver,
            phrase_reader=phrase_reader, clause_reader=clause_reader, sentence_reader=sentence_reader,
            graph_processor=self.graph_processor,
        )

    def tokenize_prompt(self, prompt: UserPrompt) -> Subject:
        return self.tokenizer.tokenize_prompt(prompt)

    def read_sentence(self, text: str, *, trace: Optional[List[dict]] = None) -> Sentence:
        """Reads `text` as exactly one sentence (spec 14.3) -- delegates
        to the shared SentenceReader via reading_context, never
        re-implements sequencing here (spec 9). `trace`, when a list is
        passed, is filled with one record per token position describing
        every phrase type PhraseReader.read() attempted there -- see
        that method's own docstring (role/phrase_reader.py) and
        ui/sentence_reader_server.py, the one consumer of this."""
        return self.reading_context.sentence_reader.read(text, grammar=self.grammar_configurator, trace=trace)

    def read_text(self, text: str) -> List[Sentence]:
        """Splits `text` into sentences the same way tokenize_prompt's
        write path does (LinguisticLexer.split_sentences), then reads
        each one independently via the shared SentenceReader -- this
        phase has no cross-sentence discourse structure to preserve
        (spec 22's own out-of-scope list includes "discourse analysis"),
        so each Sentence.read() call stands alone."""
        raw_sentences = LinguisticLexer.split_sentences(text, self.grammar_configurator)
        return [
            self.reading_context.sentence_reader.read(sentence_text, grammar=self.grammar_configurator, sequence_number=idx)
            for idx, sentence_text in enumerate(raw_sentences)
            if sentence_text
        ]

File: prototype/src/lira/linguistics/role/linguistic_controller.ts
import type { DictionaryProcessor } from "../../vocabulary/role/dictionary_processor";
import type { LexicalRelationshipStore } from "../../vocabulary/data/lexical_relationship_store";
import type { PartOfSpeech } from "../../vocabulary/data/enums/part_of_speech";
import type { Document } from "../data/document";
import type { Paragraph } from "../data/paragraph";
import type { Sentence } from "../data/sentence";
import { LinguisticSystemPropertyTensor } from "../data/tensor";
import { ValidationOutcome } from "../data/validation_outcome";
import type { UserPrompt } from "../ui/user_prompt";
import { ClauseReader } from "./clause_reader";
import { DocumentReader } from "./document_reader";
import { GrammarConfigurator } from "./grammar_configurator";
import { GraphProcessor } from "./graph_processor";
import { LexicalEvidenceStore } from "./lexical_evidence_store";
import { LinguisticLexer } from "./lexer";
import { ParagraphReader } from "./paragraph_reader";
import { PhraseReader } from "./phrase_reader";
import { PromptTokenizer } from "./prompt_tokenizer";
import type { ReadingContext } from "./reading_context";
import { SentenceReader } from "./sentence_reader";
import { SequenceEngine } from "./sequence_engine";
import { TokenResolver } from "./token_resolver";

/** LinguisticController: wires the rest of the Linguistics Service
 * together, the same role DomainController plays for Domain --
 * GraphProcessor, PromptTokenizer, LinguisticLexer,
 * ClauseSegmentationUtility, GrammarConfigurator, and the read-path
 * services (SequenceEngine, TokenResolver, PhraseReader, ClauseReader,
 * SentenceReader), bundled once into a ReadingContext every read call
 * shares.
 *
 * Ported from linguistics/role/linguistic_controller.py. */
export class LinguisticController {
  readonly grammarConfigurator: GrammarConfigurator;
  readonly tensor: LinguisticSystemPropertyTensor;
  readonly graphProcessor: GraphProcessor;
  readonly tokenizer: PromptTokenizer;
  readonly lexicalRelationships?: LexicalRelationshipStore;
  readonly readingContext: ReadingContext;
  /** spec 15-24's Proposed learned lexical transition evidence store --
   * always constructed (never undefined), so `recordObservedReading`/
   * `learningObservationCount` are always meaningful to call, but only
   * ever *read from* during scoring if a caller actually threads it
   * into SequenceEngine, which this constructor always does. Whether
   * evidence ever gets recorded into it is entirely up to callers of
   * `recordObservedReading` -- a controller nobody ever calls that on
   * behaves exactly as if learning didn't exist. */
  readonly evidenceStore: LexicalEvidenceStore;

  /** dictionaryProcessor: Vocabulary owns the lexicon (Rule 17);
   * Linguistics resolves tokens through it rather than keeping its own
   * copy (typically domain.vocabulary.dictionaryProcessor).
   * lexicalRelationships: plumbed through for Phase 2 (real
   * morphological agreement scoring); no Phase 1 reader consults it. */
  constructor(
    dictionaryProcessor: DictionaryProcessor,
    useClauseSegmentation = true,
    lexicalRelationships?: LexicalRelationshipStore,
  ) {
    this.grammarConfigurator = new GrammarConfigurator();
    // A typo in a rule table fails here, at construction time, not
    // mid-parse.
    this.grammarConfigurator.validateAgainstVocabulary();

    this.tensor = new LinguisticSystemPropertyTensor(); // persistent, canonical store for every unit's numeric fields (Rule 14)
    this.graphProcessor = new GraphProcessor(dictionaryProcessor, this.grammarConfigurator, this.tensor, useClauseSegmentation);
    this.tokenizer = new PromptTokenizer(this.graphProcessor);
    this.lexicalRelationships = lexicalRelationships;

    this.evidenceStore = new LexicalEvidenceStore();
    const sequenceEngine = new SequenceEngine(this.grammarConfigurator, undefined, this.evidenceStore);
    const tokenResolver = new TokenResolver(this.graphProcessor);
    const phraseReader = new PhraseReader(sequenceEngine, this.graphProcessor, this.grammarConfigurator);
    const clauseReader = new ClauseReader(phraseReader, sequenceEngine, this.grammarConfigurator);
    const sentenceReader = new SentenceReader(clauseReader, tokenResolver, sequenceEngine, this.grammarConfigurator);
    const paragraphReader = new ParagraphReader(sentenceReader, this.grammarConfigurator);
    const documentReader = new DocumentReader(paragraphReader, this.grammarConfigurator);

    // Built once, held for the controller's lifetime -- every
    // readPhrase()/readClause()/readSentence()/readParagraph()/
    // readDocument() call reaches these same shared services through
    // this bundle.
    this.readingContext = {
      grammar: this.grammarConfigurator, sequenceEngine, tokenResolver,
      phraseReader, clauseReader, sentenceReader, paragraphReader, documentReader,
      graphProcessor: this.graphProcessor,
    };
  }

  tokenizePrompt(prompt: UserPrompt): Document {
    return this.tokenizer.tokenizePrompt(prompt);
  }

  /** Reads `text` as exactly one sentence (spec 14.3) -- delegates to
   * the shared SentenceReader via readingContext, never re-implements
   * sequencing here (spec 9). */
  readSentence(text: string, trace?: unknown[]): Sentence {
    return this.readingContext.sentenceReader.read(text, { grammar: this.grammarConfigurator, trace });
  }

  /** Splits `text` into sentences the same way tokenizePrompt's write
   * path does (LinguisticLexer.splitSentences), then reads each one
   * independently via the shared SentenceReader -- this phase has no
   * cross-sentence discourse structure to preserve, so each read call
   * stands alone. */
  readText(text: string): Sentence[] {
    const rawSentences = LinguisticLexer.splitSentences(text, this.grammarConfigurator);
    return rawSentences
      .filter((sentenceText) => sentenceText)
      .map((sentenceText, idx) => this.readingContext.sentenceReader.read(sentenceText, { grammar: this.grammarConfigurator, sequenceNumber: idx }));
  }

  /** Reads `text` as one Paragraph -- delegates to the shared
   * ParagraphReader via readingContext, never re-implements sentence
   * splitting or sequencing here (spec 9, extended one level up). */
  readParagraph(text: string): Paragraph {
    return this.readingContext.paragraphReader.read(text, { grammar: this.grammarConfigurator });
  }

  /** Reads `text` as one Document -- classifies each line as a Heading
   * or a Paragraph (heading.ts's own matchHeadingLine) and delegates to
   * the shared DocumentReader via readingContext. This is the top of
   * the "Document, Heading, Paragraph, Sentence, Phrase, Word" reading
   * hierarchy: DocumentReader -> ParagraphReader -> SentenceReader ->
   * ClauseReader/PhraseReader -> SequenceEngine, each level delegating
   * to exactly the one below it. */
  readDocument(text: string): Document {
    return this.readingContext.documentReader.read(text, { grammar: this.grammarConfigurator });
  }

  /** spec 17's "Validated observation => lexical evidence increases" --
   * the *only* place this controller ever writes to `evidenceStore`.
   * Gated on the whole sentence's own ValidationOutcome, not each
   * phrase's individually: clause_reader.ts's own validate() already
   * defines a clause's validation as the worst of its phrases', so a
   * VALID sentence's independent clause -- and therefore every phrase
   * in it -- is never worse than VALID; the per-phrase check below is
   * cheap belt-and-braces, not load-bearing. A degenerate/unresolved
   * phrase (`phrase.words.length === 0`, `unreadablePhrase`) has no
   * real transitions to record and is skipped. Returns how many
   * transitions this call actually recorded (0 if the sentence didn't
   * validate) -- purely informational, for a caller to report back to
   * its own UI. */
  recordObservedReading(sentence: Sentence): number {
    if (sentence.validation !== ValidationOutcome.VALID) return 0;
    let recorded = 0;
    for (const clause of sentence.clauses) {
      for (const phrase of clause.phrases) {
        if (phrase.phraseType === undefined || phrase.words.length === 0 || phrase.validation !== ValidationOutcome.VALID) continue;
        let fromState: PartOfSpeech | undefined;
        for (const word of phrase.words) {
          this.evidenceStore.record(phrase.phraseType, fromState, word.partOfSpeech);
          fromState = word.partOfSpeech;
          recorded += 1;
        }
      }
    }
    return recorded;
  }
}
